pkg/faketikv/client.go

```diff
@@ -25,5 +39,273 @@ type Client interface {
+			return
+		}
+		select {
+		case c.receiveRegionHeartbeatCh <- resp:
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *client) reportRegionHeartbeat(ctx context.Context, stream pdpb.PD_RegionHeartbeatClient, errCh chan error, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case region := <-c.reportRegionHeartbeatCh:
+			request := &pdpb.RegionHeartbeatRequest{
+				Header:          c.requestHeader(),
+				Region:          region.Region,
+				Leader:          region.Leader,
+				DownPeers:       region.DownPeers,
+				PendingPeers:    region.PendingPeers,
+				BytesWritten:    region.WrittenBytes,
+				BytesRead:       region.ReadBytes,
+				ApproximateSize: uint64(region.ApproximateSize),
+			}
+			err := stream.Send(request)
+			if err != nil {
+				errCh <- err
+				log.Errorf("[store %d][pd] report regionHeartbeat error: %v", c.tag, err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *client) Close() {
+	c.cancel()
+	c.wg.Wait()
+
+	if err := c.clientConn.Close(); err != nil {
+		log.Errorf("[store %d][pd] failed close grpc clientConn: %v", c.tag, err)
+	}
+}
+
+func (c *client) GetClusterID(context.Context) uint64 {
+	return c.clusterID
+}
+
+func (c *client) AllocID(ctx context.Context) (uint64, error) {
+	ctx, cancel := context.WithTimeout(ctx, pdTimeout)
+	resp, err := c.pdClient().AllocID(ctx, &pdpb.AllocIDRequest{
+		Header: c.requestHeader(),
+	})
+	cancel()
+	if err != nil {
+		return 0, err
+	}
+	return resp.GetId(), nil
+}
+
+func (c *client) Bootstrap(ctx context.Context, store *metapb.Store, region *metapb.Region) error {
+	ctx, cancel := context.WithTimeout(ctx, pdTimeout)
+	_, err := c.pdClient().Bootstrap(ctx, &pdpb.BootstrapRequest{
+		Header: c.requestHeader(),
+		Store:  store,
+		Region: region,
+	})
+	cancel()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *client) PutStore(ctx context.Context, store *metapb.Store) error {
+	ctx, cancel := context.WithTimeout(ctx, pdTimeout)
+	resp, err := c.pdClient().PutStore(ctx, &pdpb.PutStoreRequest{
+		Header: c.requestHeader(),
+		Store:  store,
+	})
+	cancel()
+	if err != nil {
+		return err
+	}
+	if resp.Header.GetError() != nil {
+		log.Info(resp.Header.GetError())
+		return nil
+	}
+	return nil
+}
+
+func (c *client) StoreHeartbeat(ctx context.Context, stats *pdpb.StoreStats) error {
+	ctx, cancel := context.WithTimeout(ctx, pdTimeout)
+	resp, err := c.pdClient().StoreHeartbeat(ctx, &pdpb.StoreHeartbeatRequest{
+		Header: c.requestHeader(),
+		Stats:  stats,
+	})
+	cancel()
+	if err != nil {
+		return err
+	}
+	if resp.Header.GetError() != nil {
+		log.Info(resp.Header.GetError())
+		return nil
+	}
+	return nil
+}
+
+func (c *client) RegionHeartbeat(ctx context.Context, region *core.RegionInfo) error {
+	c.reportRegionHeartbeatCh <- region
+	return nil
+}
+
+func (c *client) requestHeader() *pdpb.RequestHeader {
+	return &pdpb.RequestHeader{
+		ClusterId: c.clusterID,
+	}
+}
+
+func addrsToUrls(addr string) string {
```

Comment:
no need for multiple addrs? if so addrsToUrls -> addrToUrl

Reply:
it is for test, seem only one PD is enough.

Reply:
but I don't think we need this function, forcibly using http is fine.